From f7c8789987eea9455a4185900182d237d3020899 Mon Sep 17 00:00:00 2001
From: Pranav Vashi <neobuddy89@gmail.com>
Date: Sun, 17 Aug 2025 21:25:58 +0530
Subject: [PATCH 06/32] Launcher3: Quickspace: Improve adding/removing
 listeners

Squashed:

    From: Pranav Vashi <neobuddy89@gmail.com>
    Date: Thu, 4 Sep 2025 02:50:34 +0530
    Subject: Launcher3: QuickSpace: Fix new memory leaks

    Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>

    From: Pranav Vashi <neobuddy89@gmail.com>
    Date: Thu, 11 Sep 2025 20:38:04 +0530
    Subject: Launcher3: QuickSpace: Clear runnables on pause

    * We don't need to explicitly do this in onDestroy() since removeListener(..) will do that.

    Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>

    From: Pranav Vashi <neobuddy89@gmail.com>
    Date: Fri, 12 Sep 2025 23:12:19 +0530
    Subject: Launcher3: QuickSpace: Remove psa callbacks before scheduling posting new

    * And some other clean ups.

    Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>

Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 .../launcher3/quickspace/QuickSpaceView.java  |  73 ++++++------
 .../quickspace/QuickspaceController.java      | 108 +++++++++++-------
 2 files changed, 109 insertions(+), 72 deletions(-)

diff --git a/src/com/android/launcher3/quickspace/QuickSpaceView.java b/src/com/android/launcher3/quickspace/QuickSpaceView.java
index 7e93558630..6434e08b98 100644
--- a/src/com/android/launcher3/quickspace/QuickSpaceView.java
+++ b/src/com/android/launcher3/quickspace/QuickSpaceView.java
@@ -63,9 +63,7 @@ public class QuickSpaceView extends FrameLayout implements OnDataListener {
     public TextView mEventTitle;
 
     public boolean mIsQuickEvent;
-    public boolean mFinishedInflate;
     public boolean mWeatherAvailable;
-    public boolean mAttached;
 
     private boolean mIsAlternateStyle = false;
 
@@ -158,6 +156,7 @@ public class QuickSpaceView extends FrameLayout implements OnDataListener {
         tv.setEllipsize(TruncateAt.END);
         final float textWidth = tv.getPaint().measureText(tv.getText().toString());
         tv.post(() -> {
+            if (!tv.isAttachedToWindow()) return;
             android.text.Layout layout = tv.getLayout();
             if (layout != null && layout.getEllipsizedWidth() < textWidth) {
                 tv.setEllipsize(TruncateAt.MARQUEE);
@@ -223,15 +222,40 @@ public class QuickSpaceView extends FrameLayout implements OnDataListener {
         }
     }
 
-    private void prepareLayout(boolean useAlternativeQuickspaceUI) {
-        mIsAlternateStyle = useAlternativeQuickspaceUI;
-        int indexOfChild = indexOfChild(mQuickspaceContent);
-        removeView(mQuickspaceContent);
-        if (useAlternativeQuickspaceUI) {
-            addView(LayoutInflater.from(getContext()).inflate(R.layout.quickspace_alternate_double, this, false), indexOfChild);
-        } else {
-            addView(LayoutInflater.from(getContext()).inflate(R.layout.quickspace_doubleline, this, false), indexOfChild);
+    private void clearOldViewState() {
+        View[] vs = new View[]{ mBubbleTextView, mEventTitle, mEventTitleSub, mEventTitleSubColored,
+                mNowPlayingIcon, mEventSubIcon, mWeatherContentSub, mWeatherIconSub, mWeatherTempSub };
+        for (View v : vs) if (v != null) {
+            v.animate().cancel();
+            v.setOnClickListener(null);
+            if (v instanceof ImageView) {
+                ImageView iv = (ImageView) v;
+                iv.setImageDrawable(null);
+                iv.setImageBitmap(null);
+                iv.setBackground(null);
+            } else if (v instanceof TextView) {
+                TextView tv = (TextView) v;
+                tv.setSelected(false);
+                tv.setEllipsize(null);
+                tv.setCompoundDrawablesWithIntrinsicBounds(null, null, null, null);
+                tv.setBackground(null);
+            } else {
+                v.setBackground(null);
+            }
         }
+    }
+
+    private void prepareLayout(boolean alt) {
+        mIsAlternateStyle = alt;
+        int insertIndex = (mQuickspaceContent != null) ? indexOfChild(mQuickspaceContent) : -1;
+        if (mQuickspaceContent != null) {
+            clearOldViewState();
+            removeView(mQuickspaceContent);
+        }
+        addView(LayoutInflater.from(getContext()).inflate(
+                alt ? R.layout.quickspace_alternate_double : R.layout.quickspace_doubleline,
+                this, false),
+                insertIndex < 0 ? -1 : insertIndex);
 
         loadViews();
         getQuickSpaceView();
@@ -276,19 +300,18 @@ public class QuickSpaceView extends FrameLayout implements OnDataListener {
     @Override
     public void onAttachedToWindow() {
         super.onAttachedToWindow();
-        if (mAttached)
-            return;
-
-        mAttached = true;
+        if (mController != null) {
+            mController.addListener(this);
+        }
     }
 
     @Override
     public void onDetachedFromWindow() {
+        clearOldViewState();
         super.onDetachedFromWindow();
-        if (!mAttached)
-            return;
-
-        mAttached = false;
+        if (mController != null) {
+            mController.removeListener(this);
+        }
     }
 
     public boolean isPackageEnabled(String pkgName, Context context) {
@@ -303,7 +326,6 @@ public class QuickSpaceView extends FrameLayout implements OnDataListener {
     public void onFinishInflate() {
         super.onFinishInflate();
         loadViews();
-        mFinishedInflate = true;
         mBubbleTextView = findViewById(R.id.dummyBubbleTextView);
         mBubbleTextView.setTag(new ItemInfo() {
             @Override
@@ -312,16 +334,6 @@ public class QuickSpaceView extends FrameLayout implements OnDataListener {
             }
         });
         mBubbleTextView.setContentDescription("");
-        if (isAttachedToWindow()) {
-            if (mController != null) {
-                mController.addListener(this);
-            }
-        }
-    }
-
-    @Override
-    public void onLayout(boolean b, int n, int n2, int n3, int n4) {
-        super.onLayout(b, n, n2, n3, n4);
     }
 
     public void onPause() {
@@ -329,9 +341,6 @@ public class QuickSpaceView extends FrameLayout implements OnDataListener {
     }
 
     public void onResume() {
-        if (mController != null && mFinishedInflate) {
-            mController.addListener(this);
-        }
         mController.onResume();
     }
 
diff --git a/src/com/android/launcher3/quickspace/QuickspaceController.java b/src/com/android/launcher3/quickspace/QuickspaceController.java
index 20e5b06fef..d78ff082b2 100644
--- a/src/com/android/launcher3/quickspace/QuickspaceController.java
+++ b/src/com/android/launcher3/quickspace/QuickspaceController.java
@@ -58,16 +58,16 @@ public class QuickspaceController implements OmniJawsClient.OmniJawsObserver,
     private OmniJawsClient mWeatherClient;
     private OmniJawsClient.WeatherInfo mWeatherInfo;
     private Drawable mConditionImage;
+    private boolean mOmniRegistered = false;
 
     private static final long PSA_UPDATE_DELAY_MS = 3 * 60 * 1000;
 
     private final Handler mHandler = MAIN_EXECUTOR.getHandler();
-    private final Runnable mPsaRunnable;
 
-    private Runnable mOnDataUpdatedRunnable = new Runnable() {
+    private final Runnable mOnDataUpdatedRunnable = new Runnable() {
             @Override
             public void run() {
-                for (OnDataListener list : mListeners) {
+                for (OnDataListener list : new ArrayList<>(mListeners)) {
                     list.onDataUpdated();
                 }
             }
@@ -77,6 +77,7 @@ public class QuickspaceController implements OmniJawsClient.OmniJawsObserver,
             @Override
             public void run() {
                 try {
+                    if (mWeatherClient == null) return;
                     mWeatherClient.queryWeather(mContext);
                     mWeatherInfo = mWeatherClient.getWeatherInfo();
                     if (mWeatherInfo != null) {
@@ -89,6 +90,17 @@ public class QuickspaceController implements OmniJawsClient.OmniJawsObserver,
             }
         };
 
+    private Runnable mPsaRunnable = new Runnable() {
+            @Override
+            public void run() {
+                mHandler.removeCallbacks(this);
+                if (mEventsController == null) return;
+                mEventsController.updatePsonality();
+                mHandler.postDelayed(this, PSA_UPDATE_DELAY_MS);
+                notifyListeners();
+            }
+        };
+
     public interface OnDataListener {
         void onDataUpdated();
     }
@@ -97,40 +109,53 @@ public class QuickspaceController implements OmniJawsClient.OmniJawsObserver,
         mContext = context;
         mConditionMap = initializeConditionMap();
         mEventsController = new QuickEventsController(context);
-        mWeatherClient = OmniJawsClient.get();
-
-        mPsaRunnable = new Runnable() {
-            @Override
-            public void run() {
-                if (mEventsController != null) {
-                    mEventsController.updatePsonality();
-                    notifyListeners();
-                }
-                mHandler.postDelayed(this, PSA_UPDATE_DELAY_MS);
-            }
-        };
     }
 
     private void addWeatherProvider() {
         if (!LauncherPrefs.SHOW_QUICKSPACE_WEATHER.get(mContext)) return;
-        mWeatherClient.addObserver(mContext, this);
+        if (mWeatherClient == null) mWeatherClient = OmniJawsClient.get();
+        if (!mOmniRegistered) {
+            mWeatherClient.addObserver(mContext, this);
+            mOmniRegistered = true;
+        }
         queryAndUpdateWeather();
     }
 
     public void addListener(OnDataListener listener) {
-        mListeners.add(listener);
-        addWeatherProvider();
-        MSMHProxy.INSTANCE(mContext).addMediaMetadataListener(this);
-        mEventsController.initQuickEvents();
-        mHandler.post(mPsaRunnable);
+        if (listener == null) return;
+        boolean wasEmpty = mListeners.isEmpty();
+        if (!mListeners.contains(listener)) {
+            mListeners.add(listener);
+        }
+        if (wasEmpty) {
+            addWeatherProvider();
+            registerMediaController();
+            mEventsController.initQuickEvents();
+            mHandler.post(mPsaRunnable);
+        }
         listener.onDataUpdated();
     }
 
-    private void removeListener(OnDataListener listener) {
-        if (mWeatherClient != null) {
+    private void removeOmniIfRegistered() {
+        if (mOmniRegistered && mWeatherClient != null) {
             mWeatherClient.removeObserver(mContext, this);
+            mOmniRegistered = false;
         }
+        mWeatherClient = null;
+        mWeatherInfo = null;
+        mConditionImage = null;
+    }
+
+    public void removeListener(OnDataListener listener) {
+        if (listener == null) return;
         mListeners.remove(listener);
+        if (mListeners.isEmpty()) {
+            removeOmniIfRegistered();
+            unregisterMediaController();
+            mHandler.removeCallbacks(mPsaRunnable);
+            mHandler.removeCallbacks(mWeatherRunnable);
+            mHandler.removeCallbacks(mOnDataUpdatedRunnable);
+        }
     }
 
     public boolean isQuickEvent() {
@@ -214,27 +239,27 @@ public class QuickspaceController implements OmniJawsClient.OmniJawsObserver,
     }
 
     public void onPause() {
-        cancelListeners();
+        unregisterMediaController();
+        mHandler.removeCallbacks(mPsaRunnable);
+        mHandler.removeCallbacks(mWeatherRunnable);
+        mHandler.removeCallbacks(mOnDataUpdatedRunnable);
     }
 
     public void onResume() {
+        registerMediaController();
         updateMediaController();
+        mHandler.post(mPsaRunnable);
         notifyListeners();
     }
 
-    private void cancelListeners() {
+    public void onDestroy() {
+        unregisterMediaController();
+        mHandler.removeCallbacks(mPsaRunnable);
+        mHandler.removeCallbacks(mWeatherRunnable);
+        mHandler.removeCallbacks(mOnDataUpdatedRunnable);
         for (OnDataListener listener : new ArrayList<>(mListeners)) {
             removeListener(listener);
         }
-        unregisterMediaController();
-    }
-
-    public void onDestroy() {
-        cancelListeners();
-        mHandler.removeCallbacks(mPsaRunnable);
-        mWeatherClient = null;
-        mWeatherInfo = null;
-        mConditionImage = null;
     }
 
     @Override
@@ -265,14 +290,17 @@ public class QuickspaceController implements OmniJawsClient.OmniJawsObserver,
         mHandler.post(mOnDataUpdatedRunnable);
     }
 
+    private void registerMediaController() {
+        MSMHProxy.INSTANCE(mContext).addMediaMetadataListener(this);
+    }
+
     private void unregisterMediaController() {
         MSMHProxy.INSTANCE(mContext).removeMediaMetadataListener(this);
     }
 
-    private void updateMediaController() {
+    private boolean updateMediaController() {
         if (!LauncherPrefs.SHOW_QUICKSPACE_NOWPLAYING.get(mContext)) {
-            unregisterMediaController();
-            return;
+            return false;
         }
         MediaMetadata mediaMetadata = MSMHProxy.INSTANCE(mContext).getCurrentMediaMetadata();
         boolean isPlaying = MSMHProxy.INSTANCE(mContext).isMediaPlaying();
@@ -280,16 +308,16 @@ public class QuickspaceController implements OmniJawsClient.OmniJawsObserver,
         String trackTitle = isPlaying && mediaMetadata != null ? mediaMetadata.getString(MediaMetadata.METADATA_KEY_TITLE) : "";
         mEventsController.setMediaInfo(trackTitle, trackArtist, isPlaying);
         mEventsController.updateQuickEvents();
-        notifyListeners();
+        return true;
     }
 
     @Override
     public void onMediaMetadataChanged() {
-        updateMediaController();
+        if (updateMediaController()) notifyListeners();
     }
 
     @Override
     public void onPlaybackStateChanged() {
-        updateMediaController();
+        if (updateMediaController()) notifyListeners();
     }
 }
-- 
2.54.0

